feat: rust dockerfile example (#199)
CerebriumAI/documentation: cerebrium/container-images/custom-dockerfiles.mdx

---
title: "Containerized Runtimes with Custom Dockerfiles"
description: "Run generic containerized applications on Cerebrium with your own Dockerfile"
---

> For even greater flexibility, you are able to bring your own, already functional containerized applications to Cerebrium! You can build anything from standard Python to
> compiled Rust, as long as you supply a `Dockerfile` to build your Cerebrium app with.

Reviewer: ```suggestion
For enhanced flexibility, users can bring existing containerized applications to Cerebrium. These can range from standard Python applications to compiled Rust binaries, provided a functional Dockerfile is supplied to build the application.
```

## Building Dockerized Python applications

> Suppose we have a simple FastAPI server we have already dockerized.

Reviewer: ```suggestion
This example demonstrates a simple FastAPI server that has been containerized:
```

```python
from fastapi import FastAPI

app = FastAPI()

@app.post("/hello")
def hello():
    return {"message": "Hello Cerebrium!"}

@app.get("/health")
def health():
    return "Ok"
```

> Since this application is dockerized, we have a simple, previously defined `Dockerfile`:

Reviewer: ```suggestion
The application is built using the following Dockerfile:
```

```dockerfile
FROM python:3.12-bookworm
COPY . .
RUN pip install uvicorn fastapi
EXPOSE 5000
CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "5000"]
```

This is just a regular Dockerfile.
However, it does include a `CMD` clause so that the container can be run.
We can leverage this Dockerfile in Cerebrium by using it directly to build this application.

Configure this application in `cerebrium.toml` by adding a custom runtime section with a `dockerfile_path` parameter:

```toml
[cerebrium.runtime.custom]
port = 5000
healthcheck_endpoint = "/health"
dockerfile_path = "./Dockerfile"
```

The configuration requires three key parameters:

> - `port`: The port your server listens on
> - `healthcheck_endpoint`: The endpoint that confirms server health. If empty, will default to a TCP ping of your specified port
> - `dockerfile_path`: The relative path of the project Dockerfile used to build the application

Reviewer: ```suggestion
- `port`: The port the server listens on.
- `healthcheck_endpoint`: The endpoint used to confirm server health. If unspecified, defaults to a TCP ping on the configured port.
- `dockerfile_path`: The relative path to the Dockerfile used to build the application.
```

If your Dockerfile does not contain a `CMD` clause, you should also specify `entrypoint` in your `cerebrium.toml`:

```toml
[cerebrium.runtime.custom]
entrypoint = ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "5000"]
...
```

<Info>

>   If a dockerfile path is specified, you must take care to install all the
>   dependencies your app needs in the Dockerfile, as well as run any commands.
>   All dependencies specified under `cerebrium.dependencies.*` will be ignored,
>   and `cerebrium.deployment.shell_commands` &
>   `cerebrium.deployment.pre_build_commands` will not be run.

Reviewer: ```suggestion
When specifying a dockerfile_path, ensure all dependencies and necessary commands are installed and executed within the Dockerfile. Dependencies listed under `cerebrium.dependencies.*`, as well as `cerebrium.deployment.shell_commands` and `cerebrium.deployment.pre_build_commands`, will be ignored.
```

</Info>

## Building Generic Dockerized applications

> So long as you supply a Dockerfile, you are not limited to just Python!
> Here's an equivalent Rust based API server in the Axum Framework to demonstrate how this works:

Reviewer: ```suggestion
Cerebrium supports applications in languages other than Python, provided a Dockerfile is supplied. The following example demonstrates a Rust-based API server using the Axum framework:
```

---
title: "Containerized Runtimes with Custom Dockerfiles"
description: "Run generic containerized applications on Cerebrium using custom Dockerfiles."
---

For enhanced flexibility, users can bring existing containerized applications
to Cerebrium. These can range from standard Python applications to
compiled Rust binaries, provided a functional Dockerfile is supplied
to build the application.

> 

Reviewer 2: May be worth mentioning how this benefits users (e.g configure once, run everywhere, local -> Deployed env is simpler, ensuring consistency between envs) - all the things users care about.

This example demonstrates a simple FastAPI server that has been containerized:

```python
from fastapi import FastAPI

app = FastAPI()

@app.post("/hello")
def hello():
    return {"message": "Hello Cerebrium!"}

@app.get("/health")
def health():
    return "Ok"
```

The application is built using the following Dockerfile:

```dockerfile
FROM python:3.12-bookworm
COPY . .
RUN pip install uvicorn fastapi
EXPOSE 5000
CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "5000"]
```

This is a standard Dockerfile with a CMD clause that specifies how
to run the container. Use this Dockerfile in Cerebrium by referencing
it directly in the configuration file.

Update cerebrium.toml to include a custom runtime section with the `dockerfile_path` parameter:

```toml
[cerebrium.runtime.custom]
port = 5000
healthcheck_endpoint = "/health"
dockerfile_path = "./Dockerfile"
```

The configuration requires three key parameters:

- `port`: The port the server listens on.
- `healthcheck_endpoint`: The endpoint used to confirm server health. If unspecified, defaults to a TCP ping on the configured port.
- `dockerfile_path`: The relative path to the Dockerfile used to build the application.

> If your Dockerfile does not contain a `CMD` clause, you should also specify `entrypoint` in your `cerebrium.toml`:

Reviewer 2: This may be bastardising the actual meaning you intended, so please adjust accordingly. More so that we shouldn't use personal pronouns

```toml
[cerebrium.runtime.custom]
entrypoint = ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "5000"]
...
```

<Info>

>   When specifying a dockerfile_path, ensure all dependencies and necessary
>   commands are installed and executed within the Dockerfile. Dependencies listed
>   under `cerebrium.dependencies.*`, as well as

Reviewer 2: ```suggestion
  When specifying a dockerfile_path, all dependencies and necessary
  commands should be installed and executed within the Dockerfile. Dependencies listed
  under `cerebrium.dependencies.*`, as well as
```

Less instruction, more suggestion :P

> <Info>
>   When specifying a dockerfile_path, ensure all dependencies and necessary
>   commands are installed and executed within the Dockerfile. Dependencies listed
>   under `cerebrium.dependencies.*`, as well as
>   `cerebrium.deployment.shell_commands` and
>   `cerebrium.deployment.pre_build_commands`, will be ignored.
> </Info>

Reviewer 2: ```suggestion
<Warning>
  When specifying a dockerfile_path, ensure all dependencies and necessary
  commands are installed and executed within the Dockerfile. Dependencies listed
  under `cerebrium.dependencies.*`, as well as
  `cerebrium.deployment.shell_commands` and
  `cerebrium.deployment.pre_build_commands`, will be ignored.
</Warning>
```

This will work better as a warning. It's pretty NB

## Building Generic Dockerized applications

Cerebrium supports applications in languages other than Python, provided
a Dockerfile is supplied. The following example demonstrates a Rust-based
API server using the Axum framework:

```rust
use axum::{
    routing::{get, post},
    Json, Router,
};
use serde_json::json;
use std::net::SocketAddr;
use tracing::Level;

async fn hello() -> Json<serde_json::Value> {
    Json(json!({ "message": "Hello Cerebrium!" }))
}

async fn health() -> &'static str {
    "Ok"
}

#[tokio::main]
async fn main() {
    let app = Router::new()
        .route("/hello", post(hello))
        .route("/health", get(health));
    let addr = SocketAddr::from(([127, 0, 0, 1], 5000));
    tracing::info!("Listening on {}", addr);
    axum::Server::bind(&addr.parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
```

> In this case, we use a multi-stage Dockerfile:

Reviewer 2: ```suggestion
In this case, a multi-stage Dockerfile is used:
```

Personal pronouns removed. May also be worth explaining why?

```dockerfile
# Build Stage
FROM rust:bookworm as builder
WORKDIR /usr/src/app
COPY Cargo.toml Cargo.lock ./
RUN cargo fetch

# Copy the rest of the application source code and Build
COPY . .
RUN cargo build --release

# Runtime Stage
FROM scratch
COPY --from=builder /usr/src/app/target/release/axum_server /app
EXPOSE 5000
CMD ["/app"]
```

> Configure this application in `cerebrium.toml` similarly to the FastAPI webserver and we are done!

Reviewer 2: ```suggestion
Similarly to the FastAPI webserver, the application should be configured in the `cerebrium.toml` file: 
```

Personal pronouns

```dockerfile
# Python base
FROM python:3.12-bookworm
RUN apt-get update && apt-get install dumb-init
RUN update-ca-certificates

# Copy source
COPY . .

# Install Python deps
RUN pip install -r requirements.txt  # contains fastapi[standard]

# Runtime
EXPOSE 8192
CMD ["dumb-init", "--", "fastapi", "run", "main.py", "--port", "8192"]
```

This is a standard Dockerfile with a CMD clause that specifies how
to run the container. Use this Dockerfile in Cerebrium by referencing
it directly in the configuration file.

Update cerebrium.toml to include a custom runtime section with the `dockerfile_path` parameter:

```toml
[cerebrium.runtime.custom]
port = 8192
healthcheck_endpoint = "/health"
dockerfile_path = "./Dockerfile"
```

The configuration requires three key parameters:

- `port`: The port the server listens on.
- `healthcheck_endpoint`: The endpoint used to confirm server health. If unspecified, defaults to a TCP ping on the configured port.
- `dockerfile_path`: The relative path to the Dockerfile used to build the application.

If your Dockerfile does not contain a `CMD` clause, you should also specify `entrypoint` in your `cerebrium.toml`:

```toml
[cerebrium.runtime.custom]

> entrypoint = ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "5000"]

Reviewer 3: the port should be 8192 for consistency
